### PR TITLE
pkb: Add DigitalOcean cloud support.

### PR DESCRIPTION
**Pull request by klausw**
Add initial support for DigitalOcean as a cloud provider.

This uses the `curl` tool to access the DigitalOcean v2 API. You must set
 up authentication before use, see README.md for instructions.

Since DigitalOcean does not provide add-on disks, droplets always use local
SSD storage as mounted on a single root partition, and AFAIK this cannot be
repartitioned. Benchmarks that request a data directory via vm.GetScratchDir()
will use a subdirectory of the root partition.

Benchmarks that need multiple scratch disks or raw devices are **NOT**
supported and will fail to run in such configurations. This includes
*copy_throughput* in disk-to-disk mode, or *aerospike* when using
non-memory storage modes.

Usage example:

```
./pkb.py --cloud=DigitalOcean --machine_type=16gb --benchmarks=iperf
```

Working benchmarks on a 16gb VM (droplet):

* aerospike, block_storage_workload, bonnie++, cassandra_stress,
  cluster_boot, coremark, fio, hpcc, iperf, mesh_network, mongodb, netperf,
  oldisim, ping, redis, scimark2, silo, sysbench_oltp, unixbench

Test in progress:

* speccpu2006

Not yet tested:

* hadoop_terasort, hbase_ycsb

Unsupported benchmarks:

* _copy_throughput:_ this needs multiple scratch disks.
* _object_storage_service:_ DigitalOcean doesn't offer an object store.
  It should work to run the benchmark using an external object store
  provider, I haven't tested that.

### CODE REVIEW

**Comment by klausw** [comment]
I'm still looking into some of the not-yet-tested benchmarks, and the scratch disk handling is
under active discussion (see for example https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/284) and may change. Overall
it appears to be working as expected at this point, so I've put this up for feedback.

**Comment by voellm** [comment]
very cool to be this far

**Comment by klausw** [comment]
test this please.

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+    Args:
+      vm_spec: virtual_machine.BaseVirtualMachineSpec object of the vm.
+    """
+    super(DigitaloceanVirtualMachine, self).__init__(vm_spec)
+    self.droplet_id = None
+    self.max_local_disks = 1
+    self.local_disk_counter = 0
+
+  def _CreateDependencies(self):
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Shouldn't need this

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    """Create VM dependencies."""
+    pass
+
+  def _DeleteDependencies(self):
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: or this

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    """Get the instance's data."""
+    stdout, _ = util.RunCurlCommand(
+        'GET', 'droplets/%s' % self.droplet_id)
+    response = json.loads(stdout)['droplet']
+    network_interface = response['networks']['v4'][0]
+    # Not currently using private addresses. Do we need to?
+    self.internal_ip = network_interface['ip_address']
+    self.ip_address = network_interface['ip_address']
+
+  @vm_util.Retry()
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Do you need the retry around this function? If deleting a VM that doesn't exist any more is an error (non zero retcode), you'll get stuck in a loop until the retry timeout.

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+      if self.local_disk_counter > self.max_local_disks:
+        raise errors.Error('Not enough local disks.')
+
+    # Just create a local directory at the specified path, don't mount
+    # anything.
+    self.RemoteCommand('sudo mkdir -p {0} && sudo chown -R $USER:$USER {0}'
+                       .format(disk_spec.mount_point))
+    self.scratch_disks.append(digitalocean_disk.DigitaloceanDisk(disk_spec))
+
+  def GetName(self):
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Doesn't seem that this is used anywhere

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    """Get a DigitalOcean VM's unique name."""
+    return self.name
+
+  def GetLocalDisks(self):
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Shouldn't need this

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    """Returns a list of local disks on the VM.
+
+    Returns:
+      A list of strings, where each string is the absolute path to the local
+          disks on the VM (e.g. '/dev/sdb').
+    """
+    return []
+
+  def SetupLocalDisks(self):
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: or this

File: perfkitbenchmarker/digitalocean/digitalocean_disk.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  def __init__(self, disk_spec):
+    super(DigitaloceanDisk, self).__init__(disk_spec)
+
+  def Attach(self, vm):
+    pass
+
+  def Detach(self):
+    pass
+
+  def GetDevicePath(self):
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/digitalocean_disk.py]: Maybe better to raise an exception to give the user an explicit error message if this gets called for some reason

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py (hunk 1 above)
Comment (klausw) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Removed.

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py (hunk 2 above)
Comment (klausw) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Removed.

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py (hunk 3 above)
Comment (klausw) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: I've rewritten the _Create/_Delete/_Exists methods and removed the Retry wrapper from all of them, it should now use the resource level's auto retry instead. Does this look right?

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py (hunk 4 above)
Comment (klausw) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Removed.

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py (hunk 5 above)
Comment (klausw) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Removed.

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py (hunk 6 above)
Comment (klausw) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Removed.

File: perfkitbenchmarker/digitalocean/digitalocean_disk.py (hunk 7 above)
Comment (klausw) [inline perfkitbenchmarker/digitalocean/digitalocean_disk.py]: Done, throws an error now.

**Comment by klausw** [comment]
PTAL.

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    Use a directory on the main disk as a single scratch dir, but fail
+    if a test assumes multiple distinct disks, i.e. for copy benchmark.
+
+    Args:
+      disk_num: The number of the disk to mount.
+    Returns:
+      The mounted disk directory.
+    """
+
+    if disk_num > 0:
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Maybe just add this check (assert disk_num is a valid disk) to virtual_machine.BaseVirtualMachine.GetScratchDir() and delete this method

File: perfkitbenchmarker/digitalocean/util.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for working with DigitalOcean resources."""
+
+import json
+import os
+from StringIO import StringIO
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/util.py]: Is this unused?

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py (hunk 8 above)
Comment (klausw) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Done, I've put the base change in a separate commit in this pull request.

File: perfkitbenchmarker/digitalocean/util.py (hunk 9 above)
Comment (klausw) [inline perfkitbenchmarker/digitalocean/util.py]: Yes, removed.

**Comment by klausw** [comment]
PTAL, I've done some lint cleanups, reworked the disk handling (fail if not disk.STANDARD),
and activated use of private IP addresses. This is currently mandatory, it would be nice if
benchmarks could declare if they need it since not all DigitalOcean regions currently support it.
(This may not make a difference in practice since we also require the 'metadata' feature which
has the same restriction.)

File: README.md [hunk 10]
```diff
@@ -245,13 +286,28 @@ To run all benchmarks in a named set, specify the set name in the benchmarks par
...
+
+The following are some common flags used when configuring
+PerfKitBenchmaker.
+
+Flag | Notes
+-----|------
+`--help`       | see all flags
+`--cloud`      | Check where the bechmarks are run.  Choices are `GCP`, `AWS`, `Azure`, or `DigitalOcean`
+`--zone`       | This flag allows you to override the default zone. See below.
+`--benchmarks` | A comma separated list of benchmarks or benchmark sets to run such as `--benchmarks=iperf,ping` . To see the full list, run `./pkd.py --help`
```
Comment (cmccoy) [inline README.md]: `pkd` here should be `pkb`

Comment (klausw) [inline README.md]: Fixed, though this was a pre-existing condition ;-)

File: perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    self.droplet_id = response['droplet']['id']
+
+    # The freshly created droplet will be in a locked and unusable
+    # state for a while, and it cannot be deleted or modified in
+    # this state. Wait for the action to finish and check the
+    # reported result.
+    if not self._GetActionResult(response['links']['actions'][0]['id']):
+      raise errors.Resource.RetryableCreationError('Creation failed, see log.')
+
+  def _GetActionResult(self, action_id, wait_seconds=10, max_tries=90):
```
Comment (ehankland) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: module level constants for the default wait times and tries?

Comment (klausw) [inline perfkitbenchmarker/digitalocean/digitalocean_virtual_machine.py]: Done.

File: perfkitbenchmarker/virtual_machine.py [hunk 12]
```diff
@@ -495,6 +495,10 @@ def GetScratchDir(self, disk_num=0):
       The mounted disk directory.
 
     """
+    if disk_num > len(self.scratch_disks):
```
Comment (ehankland) [inline perfkitbenchmarker/virtual_machine.py]: I think this should be >=

Comment (klausw) [inline perfkitbenchmarker/virtual_machine.py]: Oops, yes. Fixed, and added tests in scratch_disk_test.py for this.
